css-view-transitions-1/Overview.bs

```diff
@@ -1015,9 +1015,17 @@ urlPrefix: https://html.spec.whatwg.org/multipage/rendering.html; type: dfn;
 		:: a {{Promise}}.
 			Initially [=a new promise=] in [=this's=] [=relevant Realm=].
 
+			Note: These promises are immediately created,
+			so rejections will cause {{unhandledrejection}}s unless they're [=mark as handled|handled=],
+			even if the getters such as {{updateCallbackDone}} are not accessed.
+
 		: <dfn>finished promise</dfn>
 		:: a {{Promise}}.
-			Initially [=a new promise=] in [=this's=] [=relevant Realm=].
+			Initially [=a new promise=] in [=this's=] [=relevant Realm=],
+				[=marked as handled=].
+
+			Note: This is [=marked as handled=] to prevent duplicate {{unhandledrejection}}s,
+			as this promise only ever rejects along with the [=update callback done promise=].
```

Comment:
This would only be a duplicate if update callback done promise is not handled. If it is handled, but the finished promise is not handled, then should that still cause an unhandledrejection?

In other words, I feel like we shouldn't try too hard to prevent duplicate unhandledrejections, since these are different promises that would both be not handled if rejected

Reply:
Imagine code like this:

```js
async function attemptTransition() {
  const transition = document.startViewTransition(callback);

  try {
    await transition.updateCallbackDone;
  } catch (err) {
    reportBadError(err);
    return;
  }

  try {
    await transition.ready;
    // Do some JS animations here
  } catch (err) {
    reportSkippedTransition(err);
    return;
  }

  try {
    await transition.finished;
    // Do some cleanup
  } catch (err) {
    reportCleanupError(err);
    return;
  }
}
```

If the update callback throws, the above code will result in two identical unhandled rejections. Although the `updateCallbackDone` promise is handled, the function exits, leaving `ready` and `finished` unhandled. That doesn't seem useful, and annoying to work around. So, my goal in this PR is to avoid unhandled rejections on `ready` and `finished` when the rejection would be the same as `updateCallbackDone`.

Note that as soon as the developer does:

```js
await transition.finished;
```

…they're creating a new unhandled promise. This PR only impacts cases where the developer doesn't do anything with the promises.

Reply:
I was imagining the case where updateCallbackDone is listened to by some app state monitor, to make sure that all of the proper dom modifications take place, whereas finished is listened to by some logging / metrics subsystem to just report failures. In that case, suppose that updateCallbackDone rejects and is handled appropriately by the app state monitor. However, the logging system doesn't handle it (maybe as an oversight) but it does not result in unhandled rejection because updateCallbackDone handled it.

I think my case is fairly contrived, but I'm just wary of handling rejections for promises that seem different enough.

WDYT?

Reply:
Yeah, I see what you're saying. But it doesn't feel that different to what would happen if both systems were listening to `updateCallbackDone`, which I think it more likely in this case (`finished` only ever rejects if `updateCallbackDone` rejects).

- The app state handles `updateCallbackDone`.
- The logging system intends to handle `updateCallbackDone`, but fails to for some code path.

There's no unhandled rejection from `updateCallbackDone`, because the app state already handled it.

I'll ask Domenic for a second opinion on this though.

Tangent: I kinda think unhandled rejections in general are a bad design. They seem to be built around single consumer.

Reply:
Domenic likes the pattern in this PR (he did a similar thing in the navigation API) https://github.com/w3c/csswg-drafts/issues/8455#issuecomment-1436832150.

But I'm not opposed to take it to the group internally, or CSSWG, if you want more eyes on it.